metrics/metrics-api/src/main/java/org/apache/james/metrics/api/MetricFactory.java

```diff
@@ -40,7 +40,7 @@
 
     default <T> Mono<T> runPublishingTimerMetric(String name, Mono<T> mono) {
         TimeMetric timer = timer(name);
-        return mono.doOnNext(ignored -> timer.stopAndPublish());
+        return mono.doAfterSuccessOrError((success, fail) -> timer.stopAndPublish());
```

Comment:
I'm not sure recording failed execution time make sense.

IMO we want to count errors and record latencies of successful operations. (the previous line is closer from this behaviour)

Reply:
I see there is another method in MetricFactory. It records both cases of success and fail:
```
    default <T> T runPublishingTimerMetric(String name, Supplier<T> operation) {
        TimeMetric timer = timer(name);
        try {
            return operation.get();
        } finally {
            timer.stopAndPublish();
        }
    }
```

Reply:
Well, maybe this  do means our `TimeMetric` does not exposes the right API.

What is done in other methods should not necessarilly be a reference IMO: We can make this evolve.

I strongly belive we should stop considering failure as a normal operation, and thus stop recording it's time.

We should have a separate metric for it.

How about :

```
public interface TimeMetric {
    String name();

    long stopAndPublish();

    void notifyFailure();
}
```

`notifyFailure` will increment a `NAME-failures` counter.

That way we could write:

```
    default <T> T runPublishingTimerMetric(String name, Supplier<T> operation) {
        TimeMetric timer = timer(name);
        try {
            return operation.get();
        } catch (Exception e) {
            timer.notifyFailure();
            throw e; // We may need either to rely on Runtimes OR trick the compiler here... Tests with sneakyThrow suppliers would be much needed.
        } finally {
            timer.stopAndPublish();
        }
    }
```

Note that we might be interested to keep track of various sources of failures but IMO that is a nice step forward.

I will create a ticket in that direction. In the meantime, I strongly support being closer to the desired behaviour in new code. That is `not recording time upon failures`.